src/flatpak-spawn.c

```diff
@@ -534,6 +534,107 @@ add_paths_to_variant (GVariantBuilder *builder, GUnixFDList *fd_list, const GStr
   return TRUE;
 }
 
+static GHashTable *opt_env = NULL;
+
+static gboolean
+opt_env_cb (G_GNUC_UNUSED const char *option_name,
+            const gchar *value,
+            G_GNUC_UNUSED gpointer data,
+            GError **error)
+{
+  g_auto(GStrv) split = g_strsplit (value, "=", 2);
+
+  g_assert (opt_env != NULL);
+
+  if (split == NULL ||
+      split[0] == NULL ||
+      split[0][0] == 0 ||
+      split[1] == NULL)
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE,
+                   "Invalid env format %s", value);
+      return FALSE;
+    }
+
+  g_hash_table_replace (opt_env,
+                        g_steal_pointer (&split[0]),
+                        g_steal_pointer (&split[1]));
+  return TRUE;
+}
+
+static gboolean
+option_env_fd_cb (G_GNUC_UNUSED const gchar *option_name,
+                  const gchar *value,
+                  G_GNUC_UNUSED gpointer data,
+                  GError **error)
+{
+  g_autofree gchar *proc_filename = NULL;
+  g_autofree gchar *env_block = NULL;
+  gsize remaining;
+  const char *p;
+  guint64 fd;
+  gchar *endptr;
+
+  g_assert (opt_env != NULL);
+
+  fd = g_ascii_strtoull (value, &endptr, 10);
+
+  if (endptr == NULL || *endptr != '\0' || fd > G_MAXINT)
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_BAD_VALUE,
+                   "Not a valid file descriptor: %s", value);
+      return FALSE;
+    }
+
+  proc_filename = g_strdup_printf ("/proc/self/fd/%d", (int) fd);
+
+  if (!g_file_get_contents (proc_filename, &env_block, &remaining, error))
```

Comment:
I know this is a bit of a weird way to spell "read fd %d until the end", but flatpak-xdg-utils doesn't use libglnx, and I didn't really want to copypaste `glnx_fd_readall_bytes()` here.